handler/src/main/java/io/netty/handler/pcap/PcapWriteHandler.java

```diff
@@ -0,0 +1,532 @@
+            } finally {
+                byteBuf.release();
+            }
+        } else {
+            this.pCapWriter = new PcapWriter(this.outputStream);
+        }
+
+        // If Channel belongs to `SocketChannel` then we're handling TCP.
+        if (ctx.channel() instanceof SocketChannel) {
+
+            // Capture correct `localAddress` and `remoteAddress`
+            if (ctx.channel().parent() instanceof ServerSocketChannel) {
+                srcAddr = (InetSocketAddress) ctx.channel().remoteAddress();
+                dstAddr = (InetSocketAddress) ctx.channel().localAddress();
+            } else {
+                srcAddr = (InetSocketAddress) ctx.channel().localAddress();
+                dstAddr = (InetSocketAddress) ctx.channel().remoteAddress();
+            }
+
+            logger.debug("Initiating Fake TCP 3-Way Handshake");
+
+            ByteBuf tcpBuf = byteBufAllocator.buffer();
+
+            try {
+                // Write SYN with Normal Source and Destination Address
+                TCPPacket.writePacket(tcpBuf, null, 0, 0, srcAddr.getPort(), dstAddr.getPort(), TCPPacket.TCPFlag.SYN);
+                completeTCPWrite(srcAddr, dstAddr, tcpBuf, byteBufAllocator, ctx);
+
+                // Write SYN+ACK with Reversed Source and Destination Address
+                TCPPacket.writePacket(tcpBuf, null, 0, 1, dstAddr.getPort(), srcAddr.getPort(), TCPPacket.TCPFlag.SYN,
+                        TCPPacket.TCPFlag.ACK);
+                completeTCPWrite(dstAddr, srcAddr, tcpBuf, byteBufAllocator, ctx);
+
+                // Write ACK with Normal Source and Destination Address
+                TCPPacket.writePacket(tcpBuf, null, 1, 1, srcAddr.getPort(), dstAddr.getPort(), TCPPacket.TCPFlag.ACK);
+                completeTCPWrite(srcAddr, dstAddr, tcpBuf, byteBufAllocator, ctx);
+            } finally {
+                tcpBuf.release();
+            }
+
+            logger.debug("Finished Fake TCP 3-Way Handshake");
+        } else if (ctx.channel() instanceof DatagramChannel) {
+            DatagramChannel datagramChannel = (DatagramChannel) ctx.channel();
+
+            // If `DatagramChannel` is connected then we can get
+            // `localAddress` and `remoteAddress` from Channel.
+            if (datagramChannel.isConnected()) {
+                srcAddr = (InetSocketAddress) ctx.channel().localAddress();
+                dstAddr = (InetSocketAddress) ctx.channel().remoteAddress();
+            }
+        }
+
+        super.channelActive(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (ctx.channel() instanceof SocketChannel) {
+            handleTCP(ctx, msg, false);
+        } else if (ctx.channel() instanceof DatagramChannel) {
+            handleUDP(ctx, msg);
+        } else {
+            logger.error("Discarding Pcap Write for Unknown Channel Type: {}", ctx.channel());
```

Comment:
I think these "Discarding…" messages, here and in TCP and UDP, should be logged as `info` I think, since `error` might lead someone to be paged at night. I'm not sure how Netty normally uses log levels, though. What do you think, @normanmaurer.

Reply:
yeah or even debug...

Reply:
Done!